src/state_circuit/memory.rs

```diff
@@ -135,27 +156,53 @@ impl<F: FieldExt, const NUM_STEPS: usize> Config<F, NUM_STEPS> {
             //      - flags[0] == 1
             //      - steps[0] == 0
 
+            let q_memory_init = {
+                let address_prev = meta.query_advice(address, Rotation::prev());
+                let address_cur = meta.query_advice(address, Rotation::cur());
+                address_cur - address_prev
+            };
+
+            let value_cur = meta.query_advice(value, Rotation::cur());
+            let flag = meta.query_advice(flag, Rotation::cur());
+            let step = meta.query_advice(step, Rotation::cur());
+
+            let check_flag_init = {
+                let one = Expression::Constant(F::one());
+                one - flag.clone()
+            };
+
             // flag == 0 or 1
             // (flag) * (1 - flag)
-            let flag = meta.query_advice(flag, Rotation::cur());
             let bool_check_flag = {
                 let one = Expression::Constant(F::one());
                 flag.clone() * (one - flag.clone())
             };
 
-            // TODO: If flag == 0 (read), and step != 0, value_prev == value_cur
+            let value_prev = meta.query_advice(value, Rotation::prev());
+            // If flag == 0 (read), and step != 0, value_prev == value_cur
+            let q_read = {
+                let one = Expression::Constant(F::one());
+                one - flag.clone()
+            };
 
             // step_prev < step_cur
```

Comment:
```suggestion
            // If address_prev == address_cur, step_prev < step_cur
```

Reply:
Cool, thanks, I will change it. BTW, shall I rename step into global_counter everywhere?

Reply:
yes, i think renaming `step` -> `global_counter` would be good.

Reply:
Cool, just renamed it. We don't need NUM_STEPS, right?